### DISCUSSION TITLE
Does conversion from TIFF use XMP-Sidecar files from darktable

### DISCUSSION THREAD

**Discussion by scholzling** [photoprism/photoprism#2376] [Q&A] [answered]
Hello,

i scan analog negatives as RAW tiff an then edit them in darktable desktop with its build in tools, such as negadoctor.
As far as i understand all info about editing is stored in the corresponding XMP-Sidecar file in the same folder.

Since Photoprism uses Darktable for conversion of tiffs to jpg, my questions is if it also uses the XMP-Sidecarfile and applies the same edits as i have done them in darktable desktop.

If yes, what do i need to consider before importing?

**Comment by lastzero** [comment]
It would work if we would convert Tiffs with Darktable. Currently not the case as we use the file format reader that comes with Go. Means the config gets (once again) more complex if we implement this.

**Reply by scholzling** [reply]
Oh ok, so only if true RAW `*.dng` files are present darktable is used? Or at which point darktable is used anyway.
If have the possiblity to scan to DNG file. Would Darktable the be used?

**Reply by lastzero** [reply]
RAWs incl DNG are converted with a Raw converter, for example Darktable. So an Option would be to classify Tiff as Raw. Maybe not the most common use case though.

**Reply by scholzling** [reply]
Yea the use case indeed is special..to classify tiff as raw needs changes in photoprism code or can i change that setting somewhere?
For me its not a big problem to scan to a DNG-File so that also could just settle it.

**Reply by lastzero** [reply]
It's static right now to reduce complexity. First time someone asks for this.

**Reply by scholzling** [reply]
Yea understandable i think it would not be worth it to implement it with tiffs. 
But importing normal dng files from cameras would be a standart usecase. And also applying edits which are done on these dng files (with compatible software like darktable) seems to also be a common usecase. As you stated above the XMP-Sidcarfiles are used if the conversion uses darktable.

So i now tried using dng files. Edited them in darktable and tried to import them in photoprism.
I reset photoprism with `docker-compose exec photoprism photoprism reset` and deleted the storage folder.

I know that the xmp-files are beeing used in conversion because the image description from the xmp-file is applied to the file and visible in photoprism.
But it seems to me that rawtherapee is primarily used for conversion of raw files. It does not apply my edits, becaus i guess it cant since they are done with darktable which uses different modules i guess.

I then reset everything again, and before importing, disabled the use of rawtherapee in the settings of photoprism.
But now no images get imported at all. When i start scanning i see my files rushing trough the small status textfield in the library/indexing tab but they dont get converted. If i re-enable rawtherapee they get converted and imported.

How is decided which tool is gonna be used and why does darktable does not get used when i disable rawtherapee? Is this an issue or am i doing something wrong.

I am using darktable 3.6.0 on my desktop and photoprism:latest

**Reply by lastzero** [reply]
Any RAW converter / editor can create and update sidecar JPEGs. I struggle to understand why this should rather be implemented in PhotoPrism, potentially using a different tool that leads to different results?

**Reply by scholzling** [reply]
Yes but those sidecar JPEGs need to be created by darktable since i think RawTherapee is not able to apply operations based on darktable operations. For example darktable contains an operation called "negadoctor" for editing negatives. In the xmp file the settings for this operation look like this:

```
<rdf:li
      darktable:num="6"
      darktable:operation="negadoctor"
      darktable:enabled="1"
      darktable:modversion="2"
      darktable:params="01000000f72f963da3a9c53c5c93323c000000000000803f0000803f0000803f0000803f0000803f0000803f0000803f0000803faaf10240cdcc4cbdbe9f9a3d000080400000403f08ac6c3f"
      darktable:multi_name=""
      darktable:multi_priority="0"
      darktable:blendop_version="11"
      darktable:blendop_params="gz14eJxjYIAACQYYOOHEgAZY0QVwggZ7CB6pfNoAAE8gGQg="/>
     <rdf:li
```

How does Photoprism decide which tool (darktable/RawTherapee) is gonna be used when it does its RAW conversion? Since in my case deactivating RawTherapee results in no import at all.

I am not asking for new functionality. In your second comment you said that darktable can be used for raw conversion.
> RAWs incl DNG are converted with a Raw converter, for example Darktable. 

and in your first comment that if darktable is used it does apply the operations stored in the xmp file.
> It would work if we would convert Tiffs with Darktable.

Since i now switched to DNGs it should therfore work.

So either:
1. I misunderstood your previous comment and photoprism does not convert the dng files based on the operations in the xmp file
2. I do something wrong
3. There is an issue with photoprism

BTW thanks for the quick replies

**Reply by scholzling** [reply]
So i dont no go and possibly just scraping the surface here but 

https://github.com/photoprism/photoprism/blob/695d6b843d008a08c4bbaaf1cfb47a498ead23d3/internal/photoprism/convert_jpeg.go#L138

is the only check i can find where a decision is made for using darktable or rawtherapee. Since darktable is activated in my config (`c.conf.DarktableEnabled()==True`) the only thing could be that *.dng is on the blacklist for darktable. Is this the case? i did not set it but maybe its set by default?

**Reply by lastzero** [reply]
You can configure blacklists for each converter. See our docs.

**Reply by scholzling** [reply]
So i checked my configuration. I did not set any blacklists and they dont seem to be set by default. So i assume that darktable will be used. From the code it should even be tried first anyway so i would not need to disable rawtherapee.

When i disable Rawtherapee darktable does not get used for conversion. And nothing gets converted. Why could that be?

My files are .dng which is raw, is not on the blacklist for darktable and darktable is not disabled. So it should run im my opinion.

This is the log i see, when i disable rawtherapee and import my originals folder
with `PHOTOPRISM_TRACE: "true"` in docker-compose.yml

```
time="2022-05-29T18:45:51Z" level=info msg="indexing originals..."

time="2022-05-29T18:45:51Z" level=info msg="classify: loading nasnet"

time="2022-05-29T18:45:54Z" level=info msg="classify: loading labels from labels.txt"

time="2022-05-29T18:45:54Z" level=info msg="index: found no .ppignore file"

time="2022-05-29T18:45:54Z" level=info msg="index: added folder /"

time="2022-05-29T18:45:54Z" level=info msg="folder: added album Analograw (path:analograw public:true)"

time="2022-05-29T18:45:54Z" level=info msg="index: added folder /analograw"

time="2022-05-29T18:45:54Z" level=info msg="folder: added album 0040 (path:analograw/0040 public:true)"

time="2022-05-29T18:45:54Z" level=info msg="index: added folder /analograw/0040"

time="2022-05-29T18:45:56Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0001.dng"

time="2022-05-29T18:45:56Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0003.dng"

time="2022-05-29T18:45:56Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0002.dng"

time="2022-05-29T18:45:56Z" level=debug msg="index: created 84d267fa0b13292dc4108932ed0b433ed6d58b57_exiftool.json"

time="2022-05-29T18:45:56Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0004.dng"

time="2022-05-29T18:45:56Z" level=debug msg="index: created 5283eaca621018e916c6ca42aae7cf82808492d0_exiftool.json"

time="2022-05-29T18:45:56Z" level=debug msg="index: created 9c2b912aef07d477064b14d2909ea8c2cea6ec11_exiftool.json"

time="2022-05-29T18:45:56Z" level=debug msg="index: created 59017546572de923a504235c469722da73abd2ee_exiftool.json"

time="2022-05-29T18:45:58Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0005.dng"

time="2022-05-29T18:45:58Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0006.dng"

time="2022-05-29T18:45:58Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0007.dng"

time="2022-05-29T18:45:58Z" level=debug msg="index: created a77f38b54825ffedc34537b9b0f5aaa23150e295_exiftool.json"

time="2022-05-29T18:45:58Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0008.dng"

time="2022-05-29T18:45:58Z" level=debug msg="index: created 39a037ec6859180bc88af3530e4058ea4d204470_exiftool.json"

time="2022-05-29T18:45:58Z" level=debug msg="index: created aff5ae73bcee9d1fad06e139607cbd47e5b921a9_exiftool.json"

time="2022-05-29T18:45:58Z" level=debug msg="index: created 63abaa7ea930683fce528ba2ac2a5187627e6b0b_exiftool.json"

time="2022-05-29T18:45:59Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0009.dng"

time="2022-05-29T18:45:59Z" level=debug msg="exiftool: extracting metadata from analograw/0040/raw0010.dng"

time="2022-05-29T18:45:59Z" level=debug msg="index: created 2ca583aeeb9cfe772a28f17109b97816b70a0737_exiftool.json"

time="2022-05-29T18:45:59Z" level=debug msg="index: created 879c7fac5f0257d4073045a610680e9b5dc48cad_exiftool.json"

time="2022-05-29T18:45:59Z" level=debug msg="faces: found no orphan markers [1.1898ms]"

time="2022-05-29T18:45:59Z" level=debug msg="markers: found no invalid references [4.0581ms]"

time="2022-05-29T18:45:59Z" level=debug msg="markers: found no missing subjects [1.1104ms]"

time="2022-05-29T18:45:59Z" level=debug msg="faces: found no ambiguous subjects [930.1µs]"

time="2022-05-29T18:45:59Z" level=debug msg="faces: found no clusters to be merged [891.3µs]"

time="2022-05-29T18:45:59Z" level=debug msg="faces: found no existing clusters"

time="2022-05-29T18:45:59Z" level=debug msg="faces: skipped clustering"

time="2022-05-29T18:45:59Z" level=debug msg="faces: found no new faces [1.9266ms]"

time="2022-05-29T18:45:59Z" level=debug msg="faces: found no unmatched markers"

time="2022-05-29T18:45:59Z" level=debug msg="faces: updated 0 markers, recognized 0 faces, 0 unknown [2.654ms]"

time="2022-05-29T18:45:59Z" level=debug msg="index: updating counts"

time="2022-05-29T18:45:59Z" level=debug msg="counts: updated 1 place [7.7831ms]"

time="2022-05-29T18:45:59Z" level=debug msg="counts: updated 0 subjects [1.296ms]"

time="2022-05-29T18:45:59Z" level=debug msg="counts: updated 0 labels [981.4µs]"

time="2022-05-29T18:45:59Z" level=debug msg="covers: updated 2 folders [4.358ms]"

time="2022-05-29T18:45:59Z" level=debug msg="removed folder:/:true:false from cache"

time="2022-05-29T18:45:59Z" level=debug msg="index: found no photos without primary file [2.0165ms]"

time="2022-05-29T18:45:59Z" level=info msg="index: found no orphan files [606.2µs]"

time="2022-05-29T18:46:00Z" level=debug msg="search: updated file index [1.0916ms]"

time="2022-05-29T18:46:00Z" level=debug msg="index: updating counts"

time="2022-05-29T18:46:00Z" level=debug msg="counts: updated 0 places [1.4611ms]"

time="2022-05-29T18:46:00Z" level=debug msg="counts: updated 0 subjects [1.5847ms]"

time="2022-05-29T18:46:00Z" level=debug msg="counts: updated 0 labels [1.1762ms]"

time="2022-05-29T18:46:00Z" level=debug msg="index: updating covers"

time="2022-05-29T18:46:00Z" level=debug msg="covers: updated 0 albums [1.552ms]"

time="2022-05-29T18:46:00Z" level=debug msg="covers: updated 0 folders [1.6737ms]"

time="2022-05-29T18:46:00Z" level=debug msg="covers: updated 0 months [1.5547ms]"

time="2022-05-29T18:46:00Z" level=debug msg="covers: updated 0 labels [2.2132ms]"

time="2022-05-29T18:46:00Z" level=debug msg="covers: updated 0 subjects [1.2692ms]"

time="2022-05-29T18:46:00Z" level=debug msg="moments: analyzing 0 photos and 0 videos, with threshold 3"

time="2022-05-29T18:46:00Z" level=debug msg="moments: not enough files"

time="2022-05-29T18:46:00Z" level=info msg="indexing completed in 8 s"

time="2022-05-29T18:46:00Z" level=debug msg="http: POST /api/v1/index (200) [8.8822001s]"
```

**Reply by scholzling** [reply]
Should i maybe open up an issue with better description and means to reproduce the behaviour? Because to me it seems photoprism should behave differently based on the documentation and code(as far as i understand). Because if i test further this discussion will get a littlebit too technical i think and the answere for the initial question is already kind of given.

Or shouldnt i expect darktable to be used for dng files in my case and there is something else i am missing?

**Comment by lastzero** [comment, marked as answer]
**This discussion is really getting a bit long, so I'll make a new thread for your follow-up question.... Where do you see that DNG files are not blacklisted for Darktable?**

There are two documented commands to find out what the defaults and current settings are:

### `photoprism config`

When running `photoprism config` in a terminal (or `docker-compose exec photoprism photoprism config` when using Docker Compose), the blacklist clearly includes DNG files as this is the default due to issues with Apple ProRAW (still not completely fixed with RawTherapee, but that's a different topic):

|         Value          |                                Name                                 |
|------------------------|---------------------------------------------------------------------|
| darktable-bin          | /usr/bin/darktable-cli                                              |
| darktable-cache-path   |                                                                     |
| darktable-config-path  |                                                                     |
| darktable-blacklist    | dng,cr3                                                             |
| rawtherapee-bin        | /usr/bin/rawtherapee-cli                                            |
| rawtherapee-blacklist  |                                                                     |
| sips-bin               |                                                                     |
| ... | ... |

### `photoprism help`

`photoprism help` confirms this and also reports the default settings:

```
   --darktable-bin COMMAND                           Darktable CLI COMMAND for RAW to JPEG conversion (default: "darktable-cli") [$PHOTOPRISM_DARKTABLE_BIN]
   --darktable-blacklist EXTENSIONS                  do not use Darktable to convert files with these EXTENSIONS (default: "dng,cr3") [$PHOTOPRISM_DARKTABLE_BLACKLIST]
   --darktable-cache-path PATH                       custom Darktable cache PATH [$PHOTOPRISM_DARKTABLE_CACHE_PATH]
   --darktable-config-path PATH                      custom Darktable config PATH [$PHOTOPRISM_DARKTABLE_CONFIG_PATH]
   --rawtherapee-bin COMMAND                         RawTherapee CLI COMMAND for RAW to JPEG conversion (default: "rawtherapee-cli") [$PHOTOPRISM_RAWTHERAPEE_BIN]
   --rawtherapee-blacklist EXTENSIONS                do not use RawTherapee to convert files with these EXTENSIONS [$PHOTOPRISM_RAWTHERAPEE_BLACKLIST]
```

### Related Docs

- https://docs.photoprism.app/getting-started/docker-compose/#examples
- https://docs.photoprism.app/getting-started/docker/#examples
- https://dl.photoprism.app/docker/docker-compose.yml

**Reply by scholzling** [reply]
Ok, i am sorry.
This command `photoprism config` went over my head although i remember checking that list. 
Somehow I assumed from the config options doc-page, that there are no defaults set and also tried to understand what the code does and did not find any hardcoded values.
Also i expected to find a "No suitable converter found" Error message since thats what i thought should happen if both converters are skipped. But i did not find any error and therefore assumed it must be something else.

I now set `PHOTOPRISM_DARKTABLE_BLACKLIST: "jpg"` to overwrite the default setting and saw in th logs that darktable-cli is used. All the operations from the xmp-file seem to be applied.

So everything working as expected!

Thanks for the support.
